reporting/src/webrequest-pipeline/page-store.js

```diff
@@ -61,229 +162,263 @@ export default class PageStore {
   }
 
   /**
    * Create a new `tabContext` for the new tab
    */
   #onTabCreated = (tab) => {
-    this.#tabs.set(tab.id, makeTabContext(tab));
+    this.#pages.set(tab.id, new Page(tab));
   };
 
   /**
    * Update an existing tab or create it if we do not have a context yet.
    */
   #onTabUpdated = (tabId, info, tab) => {
-    let tabContext = this.#tabs.get(tabId);
-    if (!tabContext) {
-      tabContext = makeTabContext(tab);
-      this.#tabs.set(tabId, tabContext);
-    }
+    const page = new Page(this.#pages.get(tabId) || tab);
 
     // Update `isPrivate` and `url` if available
-    tabContext.isPrivate = tab.incognito;
-    setActive(tabContext, tab.active);
+    page.isPrivate = tab.incognito;
+    page.setActive(tab.active);
     if (info.url !== undefined) {
-      tabContext.url = info.url;
+      page.url = info.url;
     }
+
+    this.#pages.set(tabId, page);
   };
 
   /**
    * Remove tab context for `tabId`.
    */
   #onTabRemoved = (tabId) => {
-    const tabContext = this.#tabs.get(tabId);
-    if (tabContext && tabContext.state === PAGE_LOADING_STATE.COMPLETE) {
-      this.#stagePage(tabContext);
+    const serializedPage = this.#pages.get(tabId);
+    if (!serializedPage) {
+      return;
+    }
+    const page = new Page(serializedPage);
+    if (page.state === PAGE_LOADING_STATE.COMPLETE) {
+      this.#stagePage(page);
     }
-    this.#tabs.delete(tabId);
+    this.#pages.delete(tabId);
   };
 
   #onTabActivated = ({ previousTabId, tabId }) => {
     // if previousTabId is not set (e.g. on chrome), set all tabs to inactive
     // otherwise, we only have to mark the previous tab as inactive
     if (!previousTabId) {
-      for (const tab of this.#tabs.values()) {
-        setActive(tab, false);
+      for (const serializedPage of this.#pages.values()) {
+        const page = new Page(serializedPage);
+        page.setActive(false);
+        this.#pages.set(page.id, page);
       }
-    } else if (this.#tabs.has(previousTabId)) {
-      setActive(this.#tabs.get(previousTabId), false);
+    } else if (this.#pages.has(previousTabId)) {
+      const page = new Page(previousTabId);
+      page.setActive(false);
+      this.#pages.set(page.id, page);
     }
-    if (this.#tabs.has(tabId)) {
-      setActive(this.#tabs.get(tabId), true);
+
+    if (this.#pages.has(tabId)) {
+      const page = new Page(this.#pages.get(tabId));
+      page.setActive(true);
+      this.#pages.set(page.id, page);
     }
   };
 
-  #onWindowFocusChanged = (focusedWindow) => {
-    chrome.tabs.query({ active: true }, (activeTabs) => {
-      activeTabs.forEach(({ id, windowId }) => {
-        const tabContext = this.#tabs.get(id);
-        if (!tabContext) {
-          return;
-        }
-        if (windowId === focusedWindow) {
-          setActive(tabContext, true);
-        } else {
-          setActive(tabContext, false);
-        }
-      });
-    });
+  #onWindowFocusChanged = async (focusedWindowId) => {
```

Comment:
More as comment, since the previous implement does the same: having asynchronous operations here, breaks the replay-based tests when windows are changed.

Avoiding it is unfortunately not a trivial change. The internal model of the browser state that the implementation is building, can have its own mapping from windows to open tabs and update it through the tabs API. That would enable the processing of events in a synchronous way, but it needs to be kept in sync.